snakemake/script.py

```diff
@@ -535,6 +558,61 @@ def script(
                 py_exec = "python"
             # use the same Python as the running process or the one from the environment
             shell("{py_exec} {f.name:q}", bench_record=bench_record)
+        elif language == "jupyter":
+            # execute notebook
+            tmp_output = "{fname}.processed.ipynb".format(fname=f.name)
+            shell(
+                "jupyter nbconvert --execute --output {tmp_output:q} --to notebook --ExecutePreprocessor.timeout=-1 {f.name:q}",
+                bench_record=bench_record,
+                tmp_output=tmp_output,
+            )
+
+            # determine whether to save output
+            notebook_relpath = output.get("notebook_output", None)
+
+            if notebook_relpath is not None:
+                # determine output format
+                _, ext = os.path.splitext(notebook_relpath)
+                output_format = {
+                    ".ipynb": "notebook",
+                    ".html": "html",
+                    ".tex": "latex",
+                    ".pdf": "pdf",
+                    # ".slides": "slides",
+                    ".md": "markdown",
+                    ".txt": "asciidoc",
+                    ".rst": "rst",
+                    # ".py": "script",
+                }.get(ext, None)
+
+                if output_format is None:
+                    raise WorkflowError(
+                        "Invalid Jupyter Notebook output format: '{ext}'".format(
+                            ext=ext
+                        )
+                    )
+
+                # remove preamble from output
+                nb = nbformat.read(tmp_output, as_version=nbformat.NO_CONVERT)
+
+                nb["cells"].pop(0)
```

Comment:
I think for provenance reasons it would actually be good if the preamble was kept. This way, people will be able to see all the parameters used inside of the rendered notebook. Also, they would be able to simply reexecute or modify that particular step. This can be super powerful, e.g. if a core facility passes workflows as reports including the corresponding notebooks for some plots. Power users can just open the notebooks and interactively modify e.g. the styling or add annotations for their publication. Then, (ideally) they can pass them back to the core facility for re-import into the original workflow, in order to get an updated analysis.

Reply:
Good point. Do you think this is also necessary if the notebook is converted to a non-interactive format (e.g. pdf)?
Knowing about the executed parameters is probably always a good thing, but if the rendered pdf is distributed to non-pipeline users, they might be confusing.

Reply:
I think this could be handled by some kind of explanatory comment at the beginning. Another possibility would be to print the values in a structured format instead of having the cell with the pickled object in such cases.

Reply:
I like the idea of printing the values in a structured format.
I'll see how it fits the class-based implementation discussed above.